### DISCUSSION TITLE
How do I get the thermal conductivity on the other side of the gap？

### DISCUSSION THREAD

**Discussion by nuomi68** [idaholab/moose#24613] [Q&A General] [answered]
Hello,
everyone.
I want to modify the GapConductance of MOOSE, but the code I want to implement requires the thermal conductivity on the other side of the gap. What code can I use?I noticed that GapConductance can obtain the temperature at the other side of the gap. What code should I add if I want to obtain the corresponding thermal conductivity?
[GapConductance](https://mooseframework.inl.gov/docs/doxygen/modules/GapConductance_8C_source.html#l00444)
 ```
     _gap_distance = pinfo->_distance;
      _has_info = true;

      const Elem * secondary_side = pinfo->_side;
      std::vector<std::vector<Real>> & secondary_side_phi = pinfo->_side_phi;
      std::vector<dof_id_type> secondary_side_dof_indices;
      _dof_map->dof_indices(secondary_side, secondary_side_dof_indices, _temp_var->number());

      for (unsigned int i = 0; i < secondary_side_dof_indices.size(); ++i)
      {
        // The zero index is because we only have one point that the phis are evaluated at
        _gap_temp +=
            secondary_side_phi[i][0] * (*(*_serialized_solution))(secondary_side_dof_indices[i]);
      }

```

lwz

**Comment by lindsayad** [comment]
I don't think we support getting neighbor material properties in the "legacy" thermal gap heat transfer system as it is fundamentally based on an integrated boundary condition. @bwspenc or maybe @recuero can correct me if I'm wrong. Does your model require old state information? If not, then I might recommend using the newer mortar system

**Reply by nuomi68** [reply]
Thanks for your reply!

mortar is a good module that I've tried to use, but my current model doesn't have good convergence for force contact on surfaces. It's a 3D problem. So I had to make the formulation a penalty in Contact.

I have an idea that might be helpful for my problem. I first use MaterialRealAux and use an auxvariable to store the value of the thermal conductivity. Then I should be able to get the thermal conductivity value on the other side of the gap, just like I get the value of the temperature on the other side of the gap. But using the same code as the temperature, it does not work, probably because the family of the auxvariable is MONOMIAL.

Is it possible to implement my idea. If so, how can I implement it?

```
_paired_k(coupledValue("paired_thermal_conductivity")),
_paired_k_var(_quadrature ? getVar("paired_thermal_conductivity", 0) : nullptr),
_paired_k_solution(_quadrature ? &_paired_k_var->sys().currentSolution() : nullptr),
_paired_k_dof_map(_quadrature ? &_paired_k_var->sys().dofMap() : nullptr),

void
GapConductance::computeGapValues()
{
  Node * qnode = _mesh.getQuadratureNode(_current_elem, _current_side, _qp);
  PenetrationInfo * pinfo = _penetration_locator->_penetration_info[qnode->id()];
  const Elem * secondary_side = pinfo->_side;
  std::vector<std::vector<Real>> & secondary_side_phi = pinfo->_side_phi;
  std::vector<dof_id_type> secondary_side_dof_indices;
  _gap_paired_k=0.0;

  //This line cannot be executed    *************************
  _paired_k_dof_map->dof_indices(secondary_side, secondary_side_dof_indices, _paired_k_var->number());


  for (unsigned int i = 0; i < secondary_side_dof_indices.size(); ++i)
  {
   _gap_paired_k +=
        secondary_side_phi[i][0] * (*(*_paired_k_solution))(secondary_side_dof_indices[i]);
  }
}
```
Here is the error message.

```
libMesh terminating:
Assertion `s < this->n_systems()' failed.
s = 1
this->n_systems() = 0


Stack frames: 25
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*, std::ostream&)
2: libMesh::DofObject::n_comp_group(unsigned int, unsigned int) const
3: libMesh::DofMap::_dof_indices(libMesh::Elem const&, int, std::vector<unsigned long, std::allocator<unsigned long> >&, unsigned int, unsigned int, libMesh::Node const* const*, unsigned int, unsigned int, unsigned long&) const
4: libMesh::DofMap::dof_indices(libMesh::Elem const*, std::vector<unsigned long, std::allocator<unsigned long> >&, unsigned int, int) const
```

I am very eager to get your help！

lwz

**Reply by lindsayad** [reply, marked as answer]
you are querying a side element which has no degrees of freedom associated with it. You need to be querying the volumetric element, which is `pinfo->_elem`

**Reply by nuomi68** [reply]
Thank you very much for your reply to let me solve this problem, which was difficult for me. In the end, my code looks like this
```
const Elem * secondary_elem = pinfo->_elem;
_gap_paired_k=0.0;
      
_paired_k_dof_map->dof_indices(secondary_elem, secondary_side_dof_indices, _paired_k_var->number());
  //   for (unsigned int i = 0; i < secondary_side_dof_indices.size(); ++i)
  // {
  //  _gap_paired_k +=
  //       secondary_side_phi[i][0] * (*(*_paired_k_solution))(secondary_side_dof_indices[i]);
  // }
_gap_paired_k=(*(*_paired_k_solution))(secondary_side_dof_indices[0]);
```

Once again, I would like to express my sincere gratitude for your selfless help.

lwz

**Reply by lindsayad** [reply]
thank you for your gratitude @nuomi68. it's very much appreciated. Happy Mooseing!
